apollo-federation/src/sources/connect/expand/visitors/mod.rs

```diff
@@ -0,0 +1,367 @@
+                self.last_depth().unwrap_or_default(),
+                field.name().to_string(),
+            ));
+
+            Ok(())
+        }
+    }
+
+    impl GroupVisitor<SubSelection, NamedSelection> for TestVisitor<'_> {
+        fn try_get_group_for_field(
+            &self,
+            field: &NamedSelection,
+        ) -> Result<Option<SubSelection>, FederationError> {
+            Ok(field.next_subselection().cloned())
+        }
+
+        fn get_group_fields(
+            &self,
+            group: SubSelection,
+        ) -> Result<Vec<NamedSelection>, FederationError> {
+            Ok(group
+                .selections_iter()
+                .sorted_by_key(|s| s.name())
+                .cloned()
+                .collect())
+        }
+
+        fn enter_group(
+            &mut self,
+            group: SubSelection,
+        ) -> Result<Vec<NamedSelection>, FederationError> {
+            let next_depth = self.last_depth().map(|d| d + 1).unwrap_or(0);
+            self.depth_stack.push(next_depth);
+            self.get_group_fields(group)
+        }
+
+        fn exit_group(&mut self) -> Result<(), FederationError> {
+            self.depth_stack.pop().unwrap();
+            Ok(())
+        }
+    }
+
+    #[test]
+    fn it_iterates_over_empty_path() {
+        let mut visited = Vec::new();
+        let visitor = TestVisitor::new(&mut visited);
+        let (unmatched, selection) = JSONSelection::parse("").unwrap();
+        assert!(unmatched.is_empty());
+
+        visitor
+            .walk(selection.next_subselection().cloned().unwrap())
+            .unwrap();
+        assert_snapshot!(print_visited(visited), @"");
+    }
+
+    #[test]
+    fn it_iterates_over_simple_selection() {
+        let mut visited = Vec::new();
+        let visitor = TestVisitor::new(&mut visited);
+        let (unmatched, selection) = JSONSelection::parse("a b c d").unwrap();
+        assert!(unmatched.is_empty());
+
+        visitor
+            .walk(selection.next_subselection().cloned().unwrap())
+            .unwrap();
+        assert_snapshot!(print_visited(visited), @r###"
+        a
+        b
+        c
+        d
+        "###);
+    }
+
+    #[test]
+    fn it_iterates_over_aliased_selection() {
+        let mut visited = Vec::new();
+        let visitor = TestVisitor::new(&mut visited);
+        let (unmatched, selection) =
+            JSONSelection::parse("a: one b: two c: three d: four").unwrap();
+        assert!(unmatched.is_empty());
+
+        visitor
+            .walk(selection.next_subselection().cloned().unwrap())
+            .unwrap();
+        assert_snapshot!(print_visited(visited), @r###"
+        a
+        b
+        c
+        d
+        "###);
+    }
+
+    #[test]
+    fn it_iterates_over_nested_selection() {
+        let mut visited = Vec::new();
+        let visitor = TestVisitor::new(&mut visited);
+        let (unmatched, selection) = JSONSelection::parse("a { b { c { d { e } } } }").unwrap();
```

Comment:
```suggestion
        let (unmatched, selection) = JSONSelection::parse("a { b { c { d { e } } } f } g").unwrap();
```
since i had to fix the other visitor to handle this

Reply:
Done